File: src/java/src/main/java/org/wololo/flatgeobuf/geotools/FeatureCollectionConversions.java
package org.wololo.flatgeobuf.geotools;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.*;

import com.google.flatbuffers.ByteBufferUtil;
import com.google.flatbuffers.FlatBufferBuilder;
import static com.google.flatbuffers.Constants.SIZE_PREFIX_LENGTH;

import org.wololo.flatgeobuf.generated.*;

import org.geotools.data.memory.MemoryFeatureCollection;
import org.geotools.data.simple.SimpleFeatureCollection;
import org.geotools.feature.FeatureIterator;
import org.geotools.feature.simple.SimpleFeatureBuilder;
import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
import org.locationtech.jts.geom.*;
import org.opengis.feature.simple.SimpleFeature;
import org.opengis.feature.simple.SimpleFeatureType;
import org.opengis.feature.type.AttributeDescriptor;
import org.opengis.feature.type.GeometryDescriptor;
//import org.opengis.referencing.crs.CoordinateReferenceSystem;

public class FeatureCollectionConversions {
    static byte[] magicbytes = new byte[] { 0x66, 0x67, 0x62, 0x02, 0x66, 0x67, 0x62, 0x00 };

    public static void serialize(SimpleFeatureCollection featureCollection, long featuresCount,
            OutputStream outputStream) throws IOException {
        if (featuresCount == 0)
            return;

        SimpleFeatureType featureType = featureCollection.getSchema();

        List<AttributeDescriptor> types = featureType.getAttributeDescriptors();
        List<ColumnMeta> columns = new ArrayList<ColumnMeta>();

        for (int i = 0; i < types.size(); i++) {
            AttributeDescriptor ad = types.get(i);
            if (ad instanceof GeometryDescriptor) {
                // multiple geometries per feature is not supported
            } else {
                String key = ad.getLocalName();
                Class<?> binding = ad.getType().getBinding();
                ColumnMeta column = new ColumnMeta();
                column.name = key;
                if (binding.isAssignableFrom(Boolean.class))
                    column.type = ColumnType.Bool;
                else if (binding.isAssignableFrom(Integer.class))
                    column.type = ColumnType.Int;
                else if (binding.isAssignableFrom(Long.class))
                    column.type = ColumnType.Long;
                else if (binding.isAssignableFrom(Double.class))
                    column.type = ColumnType.Double;
                else if (binding.isAssignableFrom(String.class))
                    column.type = ColumnType.String;
                else
                    throw new RuntimeException("Unknown type");
                columns.add(column);
            }
        }

        //CoordinateReferenceSystem crs = featureType.getGeometryDescriptor().getCoordinateReferenceSystem();
        byte geometryType = toGeometryType(featureType.getGeometryDescriptor().getType().getBinding());
        //byte dimensions = (byte) (crs == null ? 2 : crs.getCoordinateSystem().getDimension());

        outputStream.write(magicbytes);

        HeaderMeta headerMeta = new HeaderMeta();
        headerMeta.featuresCount = featuresCount;
        headerMeta.geometryType = geometryType;
        headerMeta.columns = columns;

        byte[] headerBuffer = buildHeader(headerMeta);
        outputStream.write(headerBuffer);

        try (FeatureIterator<SimpleFeature> iterator = featureCollection.features()) {
            while (iterator.hasNext()) {
                SimpleFeature feature = iterator.next();
                byte[] featureBuffer = FeatureConversions.serialize(feature, headerMeta);
                outputStream.write(featureBuffer);
            }
        }
    }

    public static SimpleFeatureCollection deserialize(ByteBuffer bb) {
        int offset = 0;
        if (bb.get() != magicbytes[0] || bb.get() != magicbytes[1] || bb.get() != magicbytes[2]
                || bb.get() != magicbytes[3] || bb.get() != magicbytes[4] || bb.get() != magicbytes[5]
                || bb.get() != magicbytes[6] || bb.get() != magicbytes[7])
            throw new RuntimeException("Not a FlatGeobuf file");
        bb.position(offset += magicbytes.length);
        int headerSize = ByteBufferUtil.getSizePrefix(bb);
        bb.position(offset += SIZE_PREFIX_LENGTH);
        Header header = Header.getRootAsHeader(bb);
        bb.position(offset += headerSize);
        int geometryType = header.geometryType();
        Class<?> geometryClass;
        switch (geometryType) {
        case GeometryType.Point:
            geometryClass = Point.class;
            break;
        case GeometryType.MultiPoint:
            geometryClass = MultiPoint.class;
            break;
        case GeometryType.LineString:
            geometryClass = LineString.class;
            break;
        case GeometryType.MultiLineString:
            geometryClass = MultiLineString.class;
            break;
        case GeometryType.Polygon:
            geometryClass = Polygon.class;
            break;
        case GeometryType.MultiPolygon:
            geometryClass = MultiPolygon.class;
            break;
        default:
            throw new RuntimeException("Unknown geometry type");
        }

        int columnsLength = header.columnsLength();
        ArrayList<ColumnMeta> columnMetas = new ArrayList<ColumnMeta>();
        for (int i = 0; i < columnsLength; i++) {
            ColumnMeta columnMeta = new ColumnMeta();
            columnMeta.name = header.columns(i).name();
            columnMeta.type = (byte) header.columns(i).type();
            columnMetas.add(columnMeta);
        }

        HeaderMeta headerMeta = new HeaderMeta();
        headerMeta.columns = columnMetas;
        headerMeta.geometryType = (byte) geometryType;

        SimpleFeatureTypeBuilder ftb = new SimpleFeatureTypeBuilder();
        ftb.setName("testType");
        ftb.add("geometryProperty", geometryClass);
        for (ColumnMeta columnMeta : columnMetas)
            ftb.add(columnMeta.name, columnMeta.getBinding());
        SimpleFeatureType ft = ftb.buildFeatureType();
        SimpleFeatureBuilder fb = new SimpleFeatureBuilder(ft);
        MemoryFeatureCollection fc = new MemoryFeatureCollection(ft);
        while (bb.hasRemaining()) {
            int featureSize = ByteBufferUtil.getSizePrefix(bb);
            bb.position(offset += SIZE_PREFIX_LENGTH);
            Feature feature = Feature.getRootAsFeature(bb);
            bb.position(offset += featureSize);
            SimpleFeature f = FeatureConversions.deserialize(feature, fb, headerMeta);
            fc.add(f);
        }
        return fc;
    }

    private static byte toGeometryType(Class<?> geometryClass) {
        if (geometryClass.isAssignableFrom(MultiPoint.class))
            return GeometryType.MultiPoint;
        else if (geometryClass.isAssignableFrom(Point.class))
            return GeometryType.Point;
        else if (geometryClass.isAssignableFrom(MultiLineString.class))
            return GeometryType.MultiLineString;
        else if (geometryClass.isAssignableFrom(LineString.class))
            return GeometryType.LineString;
        else if (geometryClass.isAssignableFrom(MultiPolygon.class))
            return GeometryType.MultiPolygon;
        else if (geometryClass.isAssignableFrom(Polygon.class))
            return GeometryType.Polygon;
        else
            throw new RuntimeException("Unknown geometry type");
    }

    private static byte[] buildHeader(HeaderMeta headerMeta) {
        FlatBufferBuilder builder = new FlatBufferBuilder(1024);

        int[] columnsArray = headerMeta.columns.stream().mapToInt(c -> {
            int nameOffset = builder.createString(c.name);
            int type = c.type;
            return Column.createColumn(builder, nameOffset, type);
        }).toArray();
        int columnsOffset = Header.createColumnsVector(builder, columnsArray);

        Header.startHeader(builder);
        Header.addGeometryType(builder, headerMeta.geometryType);        
        Header.addIndexNodeSize(builder, 0);
        Header.addColumns(builder, columnsOffset);
        Header.addFeaturesCount(builder, headerMeta.featuresCount);
        int offset = Header.endHeader(builder);

        builder.finishSizePrefixed(offset);

        return builder.sizedByteArray();
    }
}

File: src/net/FlatGeobuf/NTS/FeatureCollectionConversions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;

using FlatBuffers;
using NetTopologySuite.Features;
using FlatGeobuf.Index;

namespace FlatGeobuf.NTS
{
    public class ColumnMeta
    {
        public string Name { get; set; }
        public ColumnType Type { get; set; }
    }

    public class LayerMeta
    {
        public string Name { get; set; }
        public GeometryType GeometryType { get; set; }
        public byte Dimensions { get; set; }
        public IList<ColumnMeta> Columns { get; set; }
    }

    public static class FeatureCollectionConversions {
        public static byte[] ToFlatGeobuf(FeatureCollection fc, GeometryType geometryType, byte dimensions = 2, IList<ColumnMeta> columns = null) {
            ulong count = (ulong) fc.Features.LongCount();

            if (count == 0)
                return new byte[0];
            
            var index = new PackedHilbertRTree(count);
            foreach (var f in fc.Features)
            {
                var b = f.Geometry.EnvelopeInternal;
                index.Add(b.MinX, b.MinY, b.MaxX, b.MaxY);
            }
            index.Finish();

            var featureFirst = fc.Features.First();

            if (columns == null && featureFirst.Attributes != null)
                columns = featureFirst.Attributes.GetNames()
                    .Select(n => new ColumnMeta() { Name = n, Type = ToColumnType(featureFirst.Attributes.GetType(n)) })
                    .ToList();

            using (var memoryStream = new MemoryStream())
            {
                using (var featuresStream = new MemoryStream())
                using (var offsetsStream = new MemoryStream())
                using (var offetsWriter = new BinaryWriter(offsetsStream))
                {
                    ulong offset = 0;
                    for (ulong i = 0; i < count; i++)
                    {
                        var feature = fc.Features[(int)index.Indices[i]];
                        var featureGeometryType = geometryType == GeometryType.Unknown ? GeometryConversions.ToGeometryType(feature.Geometry) : geometryType;
                        var buffer = FeatureConversions.ToByteBuffer(feature, featureGeometryType, dimensions, columns);
                        featuresStream.Write(buffer, 0, buffer.Length);
                        offetsWriter.Write(offset);
                        offset += (ulong) buffer.Length;
                    }
                    var header = BuildHeader(count, geometryType, columns, index);
                    memoryStream.Write(header, 0, header.Length);
                    var indexBytes = index.ToBytes();
                    memoryStream.Write(indexBytes, 0, indexBytes.Length);
                    offsetsStream.WriteTo(memoryStream);
                    featuresStream.WriteTo(memoryStream);
                }
                return memoryStream.ToArray();
            }
        }

        private static ColumnType ToColumnType(Type type) {
            switch (Type.GetTypeCode(type)) {
                case TypeCode.Byte: return ColumnType.UByte;
                case TypeCode.SByte: return ColumnType.Byte;
                case TypeCode.Boolean: return ColumnType.Bool;
                case TypeCode.Int32: return ColumnType.Int;
                case TypeCode.Int64: return ColumnType.Long;
                case TypeCode.Double: return ColumnType.Double;
                case TypeCode.String: return ColumnType.String;
                default: throw new ApplicationException("Unknown type");
            }
        }

        public static FeatureCollection FromFlatGeobuf(byte[] bytes) {
            var fc = new NetTopologySuite.Features.FeatureCollection();

            var bb = new ByteBuffer(bytes);
            
            var headerSize = ByteBufferUtil.GetSizePrefix(bb);
            bb.Position = FlatBufferConstants.SizePrefixLength;
            var header = Header.GetRootAsHeader(bb);
            
            var count = header.FeaturesCount;
            var nodeSize = header.IndexNodeSize;
            var geometryType = header.GeometryType;

            IList<ColumnMeta> columns = null;
            if (header.ColumnsLength > 0)
            {
                columns = new List<ColumnMeta>();
                for (int i = 0; i < header.ColumnsLength; i++){
                    var column = header.Columns(i).Value;
                    columns.Add(new ColumnMeta() { Name = column.Name, Type = column.Type });
                }
            }

            bb.Position += headerSize;
            
            if (nodeSize > 0) {
                var index = new PackedHilbertRTree(count, nodeSize);
                var indexData = bytes.Skip(headerSize).Take((int) index.Size).ToArray();
                index.Load(indexData);
                bb.Position += (int) index.Size + (int) count * 8;
            }

            while (bb.Position < bb.Length) {
                var featureLength = ByteBufferUtil.GetSizePrefix(bb);
                bb.Position += FlatBufferConstants.SizePrefixLength;
                var feature = FeatureConversions.FromByteBuffer(bb, geometryType, 2, columns);
                fc.Add(feature);
                bb.Position += featureLength;
            }

            return fc;
        }

        private static byte[] BuildHeader(ulong count, GeometryType geometryType, IList<ColumnMeta> columns, PackedHilbertRTree index)
        {
            var builder = new FlatBufferBuilder(4096);

            VectorOffset? columnsOffset = null;
            if (columns != null)
            {
                var columnsArray = columns
                    .Select(c => Column.CreateColumn(builder, builder.CreateString(c.Name), c.Type))
                    .ToArray();
                columnsOffset = Header.CreateColumnsVector(builder, columnsArray);
            }

            Header.StartHeader(builder);
            Header.AddGeometryType(builder, geometryType);
            if (columnsOffset.HasValue)
                Header.AddColumns(builder, columnsOffset.Value);
            if (index != null)
                Header.AddIndexNodeSize(builder, 16);
            Header.AddFeaturesCount(builder, count);
            var offset = Header.EndHeader(builder);

            builder.FinishSizePrefixed(offset.Value);

            return builder.DataBuffer.ToSizedArray();
        }
    }
}